### PR TITLE
stdenv: cross: updateAutotoolsGnuConfigScriptsHook unconditionally

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

We updateAutotoolsGnuConfigScriptsHook unconditionally in the native bootstrap; there's no reason to conditionalize it here.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Are you saying it already is an unconditional extra dep in the native bootstrap? Then I don't we need this bit at all!

File: pkgs/stdenv/cross/default.nix [hunk 1]
```diff
@@ -79,11 +79,7 @@ in lib.init bootStages ++ [
         ++ lib.optionals
              (hostPlatform.isLinux && !buildPlatform.isLinux)
              [ buildPackages.patchelf ]
-        ++ lib.optional
-             (let f = p: !p.isx86 || builtins.elem p.libc [ "musl" "wasilibc" "relibc" ] || p.isiOS || p.isGenode;
-               in f hostPlatform && !(f buildPlatform) )
-             buildPackages.updateAutotoolsGnuConfigScriptsHook
-        ;
+        ++ [ buildPackages.updateAutotoolsGnuConfigScriptsHook ];
```
Comment (Ericson2314) [inline pkgs/stdenv/cross/default.nix:82]:
```suggestion
        ;
```

**Comment by amjoseph-nixpkgs** [comment]
> Are you saying it already is an unconditional extra dep in the native bootstrap?

Not quite.  See 3b8e3c1f3de24.  It is a dependency of native `stdenv`, and therefore changes to it cause everything everywhere to rebuild.  However it is only a *direct* dependency (i.e. hooks applied) of those packages which are part of `stdenv` itself.

The point of my comment above was that this PR will not increase the number of rebuilds caused by future `gnu-config` version bumps.

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/commit/3b8e3c1f3de24472dd6dbd1fcdb3912a067bd70c#diff-bf918d7f48f6a34a4a1acb33373dee717cff2c3f5503a7ca41d90e5f289faf1eR615 doesn't that line make it part of the final stdenv / used by anything which is built with that stdenv?

**Comment by amjoseph-nixpkgs** [comment]
> [3b8e3c1#diff-bf918d7f48f6a34a4a1acb33373dee717cff2c3f5503a7ca41d90e5f289faf1eR615](https://github.com/NixOS/nixpkgs/commit/3b8e3c1f3de24472dd6dbd1fcdb3912a067bd70c#diff-bf918d7f48f6a34a4a1acb33373dee717cff2c3f5503a7ca41d90e5f289faf1eR615) doesn't that line make it part of the final stdenv / used by anything which is built with that stdenv?

You're right.

I forgot that `stdenv` doesn't copy `stdenv.mkDerivation` doesn't prepend `defaultNativeBuildInputs` to the resulting `nativeBuildInputs`.  The additional dependencies are added in `bash`, not in Nix code.

I guess they are really *propagated* native build inputs, propagated from `stdenv`.

**Comment by amjoseph-nixpkgs** [comment]
My only remaining concern is that `setup.sh` tries to be clever and filter out any `nativeBuildInputs` that can't be executed on the `buildPlatform`.

In spite of the fact that it's a shell script, Nix believes that `gnu-config` has a meaningful execution platform associated with it.

But I'm having a hard time making this happen.

I need to read `findInputs()` more closely.  Grr, I wish we weren't doing this stuff in `bash` scripts...

**Comment by amjoseph-nixpkgs** [comment]
Ah, I see.

You can never change the `buildPlatform` using `pkgs/stdenv/cross`.

That explains it.

**Pull request closed without merging**